File: SddIA/scripts/qa/cerbero_governance_react_core.py
# -*- coding: utf-8 -*-
"""Cerbero reacción RBAC ante eventos Self-Healing (Fase 4+ agnóstico)."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any

from eda_bus_utils import _write_json_atomic, load_radamanto_config


def revoked_path(repo: Path) -> Path:
    cfg = load_radamanto_config(repo)
    rel = cfg.get("revoked_entities") or ".SddIA/cerbero/revoked_entities.json"
    return repo / rel


def load_revoked(repo: Path) -> dict[str, Any]:
    path = revoked_path(repo)
    if not path.is_file():
        return {"revoked": {}, "permanent": {}}
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return {"revoked": {}, "permanent": {}}
    data.setdefault("revoked", {})
    data.setdefault("permanent", {})
    return data


def save_revoked(repo: Path, data: dict[str, Any]) -> None:
    path = revoked_path(repo)
    path.parent.mkdir(parents=True, exist_ok=True)
    _write_json_atomic(path, data)


def is_entity_revoked(repo: Path, entity_id: str) -> bool:
    data = load_revoked(repo)
    if entity_id in data.get("permanent", {}):
        return True
    return entity_id in data.get("revoked", {})


def _resolve_entity_id(payload: dict[str, Any]) -> str | None:
    eid = payload.get("entity_id") or payload.get("target_entity_id")
    if isinstance(eid, str) and eid.strip():
        return eid.strip()
    return None


def react_to_domain_event(repo: Path, event: dict[str, Any]) -> dict[str, Any]:
    event_type = event.get("event_type")
    payload = event.get("payload") or {}
    if not isinstance(payload, dict):
        return {"ok": False, "error": "payload invalido"}
    entity_id = _resolve_entity_id(payload)
    if not entity_id:
        return {"ok": False, "error": "entity_id requerido"}
    data = load_revoked(repo)

    if event_type == "Domain_Entity_Degraded":
        data["revoked"][entity_id] = {
            "since": event.get("timestamp"),
            "reason": payload.get("reason"),
            "entity_type": payload.get("entity_type"),
        }
        save_revoked(repo, data)
        return {"ok": True, "action": "revoked", "entity_id": entity_id}

    if event_type == "Domain_Entity_Restored":
        if event.get("emitter_agent") != "radamanto":
            return {"ok": False, "error": "Domain_Entity_Restored solo desde radamanto"}
        data["revoked"].pop(entity_id, None)
        save_revoked(repo, data)
        return {"ok": True, "action": "restored", "entity_id": entity_id}

    if event_type == "Domain_Entity_Deprecated":
        data["revoked"].pop(entity_id, None)
        data["permanent"][entity_id] = {
            "since": event.get("timestamp"),
            "reason": payload.get("reason"),
            "entity_type": payload.get("entity_type"),
        }
        save_revoked(repo, data)
        return {"ok": True, "action": "permanent_block", "entity_id": entity_id}

    return {"ok": False, "error": f"event_type no soportado: {event_type}"}

File: SddIA/engine/execute-process/src/engine/cerbero_governance_react_core.rs
//! Cerbero reacción RBAC (paridad `cerbero_governance_react_core.py`).

use super::fractal_bus::load_radamanto_config;
use serde_json::{json, Value};
use std::fs;
use std::path::Path;

fn write_json_atomic(path: &Path, data: &Value) -> Result<(), String> {
    super::eda_bus_topology::write_json_atomic(path, data)
}

fn revoked_path(repo: &Path) -> Result<std::path::PathBuf, String> {
    let cfg = load_radamanto_config(repo)?;
    let rel = cfg
        .get("revoked_entities")
        .and_then(|v| v.as_str())
        .unwrap_or(".SddIA/cerbero/revoked_entities.json");
    Ok(repo.join(rel.trim_start_matches("./")))
}

fn load_revoked(repo: &Path) -> Value {
    let path = revoked_path(repo).ok();
    let Some(path) = path.filter(|p| p.is_file()) else {
        return json!({"revoked": {}, "permanent": {}});
    };
    fs::read_to_string(&path)
        .ok()
        .and_then(|t| serde_json::from_str::<Value>(&t).ok())
        .map(|mut data| {
            if !data.get("revoked").and_then(|v| v.as_object()).is_some() {
                data["revoked"] = json!({});
            }
            if !data.get("permanent").and_then(|v| v.as_object()).is_some() {
                data["permanent"] = json!({});
            }
            data
        })
        .unwrap_or(json!({"revoked": {}, "permanent": {}}))
}

fn save_revoked(repo: &Path, data: &Value) -> Result<(), String> {
    let path = revoked_path(repo)?;
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent).map_err(|e| e.to_string())?;
    }
    write_json_atomic(&path, data)
}

fn resolve_entity_id(payload: &Value) -> Option<String> {
    payload
        .get("entity_id")
        .or_else(|| payload.get("target_entity_id"))
        .and_then(|v| v.as_str())
        .map(str::trim)
        .filter(|s| !s.is_empty())
        .map(str::to_string)
}

pub fn react_to_domain_event(repo: &Path, event: &Value) -> Value {
    match react_to_domain_event_inner(repo, event) {
        Ok(v) => v,
        Err(e) => json!({"ok": false, "error": e}),
    }
}

fn react_to_domain_event_inner(repo: &Path, event: &Value) -> Result<Value, String> {
    let event_type = event
        .get("event_type")
        .and_then(|v| v.as_str())
        .ok_or("event_type requerido")?;
    let payload = event.get("payload").cloned().unwrap_or(json!({}));
    if !payload.is_object() {
        return Err("payload invalido".into());
    }
    let entity_id = resolve_entity_id(&payload).ok_or("entity_id requerido")?;
    let mut data = load_revoked(repo);

    match event_type {
        "Domain_Entity_Degraded" => {
            if let Some(revoked) = data.get_mut("revoked").and_then(|v| v.as_object_mut()) {
                revoked.insert(
                    entity_id.clone(),
                    json!({
                        "since": event.get("timestamp"),
                        "reason": payload.get("reason"),
                        "entity_type": payload.get("entity_type"),
                    }),
                );
            }
            save_revoked(repo, &data)?;
            Ok(json!({"ok": true, "action": "revoked", "entity_id": entity_id}))
        }
        "Domain_Entity_Restored" => {
            if event.get("emitter_agent").and_then(|v| v.as_str()) != Some("radamanto") {
                return Err("Domain_Entity_Restored solo desde radamanto".into());
            }
            if let Some(revoked) = data.get_mut("revoked").and_then(|v| v.as_object_mut()) {
                revoked.remove(&entity_id);
            }
            save_revoked(repo, &data)?;
            Ok(json!({"ok": true, "action": "restored", "entity_id": entity_id}))
        }
        "Domain_Entity_Deprecated" => {
            if let Some(revoked) = data.get_mut("revoked").and_then(|v| v.as_object_mut()) {
                revoked.remove(&entity_id);
            }
            if let Some(permanent) = data.get_mut("permanent").and_then(|v| v.as_object_mut()) {
                permanent.insert(
                    entity_id.clone(),
                    json!({
                        "since": event.get("timestamp"),
                        "reason": payload.get("reason"),
                        "entity_type": payload.get("entity_type"),
                    }),
                );
            }
            save_revoked(repo, &data)?;
            Ok(json!({"ok": true, "action": "permanent_block", "entity_id": entity_id}))
        }
        other => Err(format!("event_type no soportado: {other}")),
    }
}
